e2e: increase the container probing test timeout

yujuhong opened the pull request:
This test runs in the parallel suite and should allow a longer timeout.

yujuhong commented:
This addresses #17656 partially.

@wojtek-t @timstclair

yujuhong commented:
cc @kubernetes/goog-testing, tests that require pulling images in a parallel environment may need a longer timeout.

timstclair commented on test/e2e/container_probe.go:
```diff
@@ -75,7 +75,7 @@ var _ = Describe("Probing container", func() {
 		expectNoError(err)
 
-		err = wait.Poll(poll, 90*time.Second, func() (bool, error) {
+		err = wait.Poll(poll, 120*time.Second, func() (bool, error) {
```
Since this test needs to run the entire poll time to pass, maybe it would be better to add a preliminary phase that waits for the pod to be Running.

yujuhong replied:
Done. The test doesn't correctly test the initial delay at all given that it also includes the pod startup time.

wojtek-t commented on test/e2e/container_probe.go:
```diff
@@ -43,7 +43,7 @@ var _ = Describe("Probing container", func() {
 		startTime := time.Now()
 
-		Expect(wait.Poll(poll, 90*time.Second, func() (bool, error) {
+		Expect(wait.Poll(poll, 120*time.Second, func() (bool, error) {
```
Can we make those timeouts even bigger?
If sometimes 90s is not enough, I really vote for something more like 240 than 120.
It's definitely better, to loose 30s from time to time, than to have flakes.

yujuhong replied:
I divided the test into two phases, and each of them is given 120s (120s for running, 120s for ready).

Note that the second test "expects" the test to time out when waiting for readiness, so I only set 90s for it.

timstclair replied:
I actually think the second phase of the first test should have a shorter timeout. The readiness should be set pretty quickly after the initial delay is up, and if it doesn't it's likely a bug (e.g. https://github.com/kubernetes/kubernetes/issues/17656). I'd suggest 30-60s, and only increasing it if that doesn't seem to be enough.

yujuhong replied:
This has to be > 30s since the initial delay is 30s. How about we settle for 60s?

timstclair replied:
Right, that's what I meant in my head... 60s sounds good.

timstclair commented on test/e2e/container_probe.go:
```diff
@@ -41,9 +41,22 @@ var _ = Describe("Probing container", func() {
+				return false, nil
+			}
+			return true, nil
+		})).NotTo(HaveOccurred(), "pod never became running")
```
Extract this to a helper function?

timstclair commented on test/e2e/container_probe.go:
```diff
@@ -41,9 +41,22 @@ var _ = Describe("Probing container", func() {
+			return true, nil
+		})).NotTo(HaveOccurred(), "pod never became running")
+
+		startTime := time.Now()
```
I'm worried this will be a little fragile now - there could be some time between when the pod became running (when the initial delay actually starts), and when we saw it as running (`time.Now()`). It would be more accurate to get the startTime from the containerStatus, but maybe that's too close to reimplementing the logic in the test?

yujuhong replied:
Yes, this is still fragile. To be more reliable and only test the probing logic, it should startTime and the not-ready->ready transition timestamp from the status as well, then compare the two. This is a bigger change than what this PR is intended though.

timstclair replied:
Here's an alternative: we know the pod was not running 1 polling period age, so just subtract `poll` form `time.Now()`

yujuhong replied:
The test polls every 2 second, and the initial delay is 30s. Subtracting 2 second doesn't really help that much IMO.

timstclair replied:
Ah, nevermind - I was thinking that the pod would be Running as soon as the container started, but since there are 2 containers that is not the case. I'm still worried this will end up flaky.

yujuhong replied:
There shouldn't be two containers in this test then. I am not arguing that this test needs more effort, but I think we should get this PR in to stop the bleeding :)

timstclair replied:
My concern is that this will just start more bleeding somewhere else. I suspect the flake caused by this will be more frequent than the flake this is fixing.

yujuhong replied:
I decided to go with the pod status timestamps. I haven't tested locally so I'll let jenkins do the testing for me. No need to review again until jenkins is green.

timstclair commented on test/e2e/container_probe.go:
```diff
@@ -41,9 +41,22 @@ var _ = Describe("Probing container", func() {
+		})).NotTo(HaveOccurred(), "pod never became running")
+
+		startTime := time.Now()
+		Expect(wait.Poll(poll, 120*time.Second, func() (bool, error) {
```
Mentioned elsewhere: I think we should consider lowering this timeout to 30-60s.

yujuhong committed:
e2e: fix container probing test

This change compare the timestamps in pod status to accurately determine
whether the initial delay has passed when the pod (container) becomes ready.
Also increase the test timeout to account for the parallel testing environment.

yujuhong commented:
@timstclair, test passed locally. PR ready for review. Note that it'd take 1 minute after the container started to report the ready pod status due to #18240.

timstclair commented:
LGTM, thanks for fixing the test!

yujuhong commented:
@k8s-bot unit test this

yujuhong commented:
Here is a brief summary of the test:

We assume the timeout is large enough for the pod to become running and ready. If this is not true, the test will fail, and we can easily see the pod state and readiness from the periodic logging. This timeout could be caused by 1) kubelet slow to start the pod; 2) kubelet slow to send out the status (due to QPS limit or other factors; 3) kubelet fails to probe the container in time. Only the third one is probing-related, but either way the test should fail.

If a pod becomes ready within the timeout, we check the timestamps in the pod status to make sure that the pod becomes ready at least 30s (initial delay) after the container starts.

We could also add a new condition to fail the test if the latency between container starting and pod becoming ready is too long.

wojtek-t commented:
I'm fine with that - although I would make timeouts even bigger.

yujuhong commented:
@fabioy (on-call), can we manually merge this PR and #18240? These two should fix the flaky container test. Thanks!

fabioy commented:
Good timing, gce-parallel just went red/yellow. Does this PR have to be merged with #18240 or can they go in separately? I'd like to try putting this one in first and see if Jenkins goes green.

yujuhong commented:
@fabioy, they can go in separately but #18240 should be merged before this one.

fabioy commented:
Well, it's still early in the day. Merging #18240 and this to help test flakiness.
